ddb-client/src/main/java/org/opensearch/remote/metadata/client/impl/DDBOpenSearchClient.java

```diff
@@ -151,6 +195,90 @@ public DDBOpenSearchClient() {}
+        AOSOpenSearchClient aosOpenSearchClient,
+        String tenantIdField,
+        ThreadPool tp
+    ) {
+        this.dynamoDbAsyncClient = dynamoDbAsyncClient;
+        this.aosOpenSearchClient = aosOpenSearchClient;
+        this.tenantIdField = tenantIdField;
+        this.threadPool = tp;
+    }
+
+    private void cacheGlobalResources() {
+        this.dynamoDbAsyncClient.listTables().thenAccept(tables -> {
+            Map<String, CompletableFuture<DescribeTableResponse>> describeTableCompletableFutureMap = getDescribeTableCompletableFutureMap(
+                tables
+            );
+            CompletableFuture.allOf(describeTableCompletableFutureMap.values().toArray(new CompletableFuture[0])).exceptionally(ex -> {
+                log.error("Failed to describe DDB tables", ex);
+                return null;
+            }).thenAccept(y -> {
+                List<String> potentialGlobalResourceTables = new ArrayList<>();
+                for (Map.Entry<String, CompletableFuture<DescribeTableResponse>> responseFuture : describeTableCompletableFutureMap
+                    .entrySet()) {
+                    DescribeTableResponse describeTableResponse = responseFuture.getValue().join();
+                    if (describeTableResponse.table().hasKeySchema()
+                        && describeTableResponse.table().keySchema().stream().anyMatch(kse -> HASH_KEY.equals(kse.attributeName()))) {
+                        potentialGlobalResourceTables.add(describeTableResponse.table().tableName());
+                    }
+                }
+                if (!potentialGlobalResourceTables.isEmpty()) {
+                    log.info("Found potential global resource tables: {}", potentialGlobalResourceTables);
+                    Map<String, CompletableFuture<QueryResponse>> fetchGlobalResourcesCompletableFutureMap = potentialGlobalResourceTables
+                        .stream()
+                        .collect(Collectors.toMap(Function.identity(), this::fetchGlobalResources));
+                    CompletableFuture.allOf(fetchGlobalResourcesCompletableFutureMap.values().toArray(new CompletableFuture[0]))
+                        .exceptionally(ex -> {
+                            log.error("Failed to execute DDB query", ex);
+                            return null;
+                        })
+                        .thenAccept(z -> {
+                            for (Map.Entry<
+                                String,
+                                CompletableFuture<QueryResponse>> responseFuture : fetchGlobalResourcesCompletableFutureMap.entrySet()) {
+                                QueryResponse queryResponse = responseFuture.getValue().join();
+                                queryResponse.items().forEach(item -> {
+                                    String id = item.get(RANGE_KEY).s();
+                                    GLOBAL_RESOURCES_CACHE.put(buildGlobalCacheKey(responseFuture.getKey(), id), item);
```

Comment:
Isn't the `getKey()` here going to be the global tenant ID?  I thought the global cache key had the index (table) name here?

Reply:
Yeah, that key is the table name, I see... and you're combining it with the ID. Got it.